docs/github-merge-strategy.md

```diff
@@ -6,6 +6,43 @@ This document describes the merge strategy used for Pull Requests within all rep
 
 Every Pull Request that is made to an InstructLab repository should meet the below requirements - certain repositories such as [taxonomy](https://github.com/instruct-lab/taxonomy) may have additional requirements.
 
+### Commit Messages and PR Description
+
+PRs should be formatted as a series of logical commits that implement the
+overall change proposed in the PR. Each commit should represent a logical step
+in the progression of the change. The commit message for each should clearly
+explain that step of the progression.
+
+Here are a few examples of PRs that contained a series of commits:
+
+- <https://github.com/instructlab/instructlab-bot/pull/125/commits>
+- <https://github.com/instructlab/instructlab/pull/994/commits>
+- <https://github.com/instructlab/instructlab/pull/951/commits>
```

Comment:
I'm sure there are even better examples out there. I'm happy to change this over time to represent some of the best examples people can find in our repos. These are a few I had dug up some weeks ago.

Reply:
Apologies if this is a personal preference nit.  Noting that many/most PRs are small in scope and (I believe) should usually only have 1 commit.  I generally think of a commit as a unit of work that you might choose to un-commit.  And breaking down work into revertable logic is a great practice.  But stating multiple commits is the general case might lead to a lot of commits for work that doesn't need to be broken up.  My ask would be to adjust the wording to not promote multiple commits quite so strongly but still indicate how to use them effectively.  If there is a fundamental disagreement on approach, I'm happy to be a chameleon.

Reply:
You're certainly not alone in this preference, and I think there's a bit of a difference between a camp of people who developed practices before github existed, and another camp who developed practices around github and maybe are less reliant on e.g. `git log`

Here's a concrete example of why I'm firmly in the "more is more" (within reason) on splitting up commits: https://github.com/instructlab/sdg/pull/74/commits

Totally trivial PR, but 4 distinctly logical changes, and being able to look at those changes (and their commit message) individually is actually really valuable - e.g. separating the reordering of the parameters from the deleting unused parameters makes it trivial to review. First you make sure the reordering didn't change anything, then you confirm that the 4 parameters being dropped aren't in use.

Reply:
Here are a few thoughts, we might consider adding:

*Splitting commits is crucial because it provides "ease of revertability." If a specific commit causes an issue, reverting a single logical unit is much easier and less disruptive than reverting an entire feature.

* Encouraging multiple PRs for large changes: contributors should be encouraged to split their work into multiple PRs if the total lines changed across commits exceed 1,000 lines. This rule of thumb helps keep changes manageable and reviews more focused.

* The approach to splitting commits and PRs can depend on the review's velocity. High velocity in review and addressing feedback leads to rapid merges. Small PRs can be reviewed and merged quickly, benefiting both contributors, who see their changes integrated faster, and maintainers, who avoid reviewing massive changes. I wish the various projects could reach that state, we are not there yet.

This strategy helps maintain the project's health and efficiency by ensuring changes are easier to manage, review, and revert if necessary.